pyomo/contrib/piecewise/triangulations.py

```diff
@@ -0,0 +1,682 @@
+                    y += 1
+                    facing = Direction.up
+                    continue
+                else:
+                    if square_parity_tlbr(x, y):
+                        add_top_left()
+                        add_bottom_right()
+                    else:
+                        add_top_right()
+                        add_bottom_left()
+                    used_squares.add((x, y))
+                    if (x, y - 1) in used_squares or y == 0:
+                        # can't keep going down so we need to turn depending
+                        # on our parity
+                        if square_parity_tlbr(x, y):
+                            x += 1
+                            facing = Direction.right
+                            continue
+                        else:
+                            x -= 1
+                            facing = Direction.left
+                            continue
+                    else:
+                        y -= 1
+                        continue
+            case Direction.up:
+                if is_turnaround(x, y):
+                    # we are always in a non-TLBR square. Take the BL of this, the BR
+                    # of the one on the left, and continue downwards one to the left
+                    add_bottom_left()
+                    x -= 1
+                    add_bottom_right()
+                    y -= 1
+                    facing = Direction.down
+                    continue
+                else:
+                    if square_parity_tlbr(x, y):
+                        add_bottom_right()
+                        add_top_left()
+                    else:
+                        add_bottom_left()
+                        add_top_right()
+                    used_squares.add((x, y))
+                    if (x, y + 1) in used_squares or y == num_pts - 1:
+                        # can't keep going up so we need to turn depending
+                        # on our parity
+                        if square_parity_tlbr(x, y):
+                            x -= 1
+                            facing = Direction.left
+                            continue
+                        else:
+                            x += 1
+                            facing = Direction.right
+                            continue
+                    else:
+                        y += 1
+                        continue
+
+
+def _get_ordered_j1_triangulation_3d(points_map, num_pts):
+    # Import these precomputed paths
+    from pyomo.contrib.piecewise.ordered_3d_j1_triangulation_data import (
+        hamiltonian_paths as simplex_pair_to_path,
+    )
+
+    # To start, we need a hamiltonian path in the grid graph of *double* cubes
+    # (2x2x2 cubes)
+    grid_hamiltonian = get_grid_hamiltonian(3, round(num_pts / 2))  # division is exact
+
+    # We always start by going from [0, 0, 0] to [0, 0, 1], so we can safely
+    # start from the -x side
+    start_data = ((-1, 0, 0), 1)
```

Comment:
The first tuple is the face, so this is saying that we start from the -x side, and the second number is the, uhh, number. We have to deal with two different simplices on each facing because if we only deal with one then some of the hamiltonian paths I brute-forced in the data file don't exist. That number can be 1 or 2.

Reply:
OK, this makes sense. Could you document that somewhere in the comments?

Reply:
Added another comment